src/controller.rs

```diff
@@ -93,8 +121,44 @@ pub fn start(
     let mut file_watcher: Option<FileWatcher> = None;
 
     'event_loop: loop {
-        select! {
-            recv(from_editor) -> msg => {
+        let never_rx = never();
+        let from_file_watcher = file_watcher
+            .as_ref()
+            .map(|fw| fw.worker.receiver())
+            .unwrap_or(&never_rx);
+        let from_pending_file_watcher = &file_watcher
+            .as_ref()
+            .and_then(
+                // If there are enqueud events, let's wait a bit for others to come in, to send
+                // them in batch.
+                |fw| {
+                    if fw.pending_file_events.is_empty() {
+                        None
+                    } else {
+                        Some(tick(Duration::from_millis(500)))
+                    }
+                },
+            )
+            .unwrap_or_else(never);
+
+        let mut sel = Select::new();
+        let server_rxs: Vec<&Receiver<ServerMessage>> = language_servers
+            .values()
+            .map(|(transport, _)| transport.from_lang_server.receiver())
+            .collect();
```

Comment:
I think `server_rxs` is loop-invariant, so we could initialize it before the loop?
I'd need to check.